File: ios/ATCTranscribe/Transcription/ATCTranscriber.swift
import Foundation
import CoreML
import WhisperKit

/// Whisper's own confidence for one decoded transmission, surfaced for the LLM confidence gate.
/// `avgLogprob` is the mean segment log-probability (closer to 0 = more confident; very negative =
/// the model was unsure); `compressionRatio` is the max over segments (high = repetitive/degenerate).
/// `noSpeechProb` is intentionally omitted — it's stubbed to 0 in this WhisperKit build.
struct ASRConfidence: Sendable, Equatable {
    var avgLogprob: Float
    var compressionRatio: Float
    /// Neutral "no signal" value for non-Whisper callers and tests (treated as confident).
    static let unknown = ASRConfidence(avgLogprob: 0, compressionRatio: 0)
}

/// One transcribed transmission plus the ASR confidence that produced it.
struct TranscriptionOutput: Sendable {
    var text: String
    var asr: ASRConfidence
    static let empty = TranscriptionOutput(text: "", asr: .unknown)
}

/// Fine-tuned Whisper inference for live ATC segments, on-device via WhisperKit
/// (CoreML on the Apple Neural Engine). Swift port of `atc_transcriber.ATCTranscriber`.
///
/// Decode policy mirrors the Python:
///  - **language pinned to `en`**, task `.transcribe` — no language auto-detect drift on
///    low-SNR radio (the Python clears `forced_decoder_ids` and pins per call).
///  - optional **airport-context prompt**, capped at ~220 tokens so decoding always has
///    room in Whisper's 448-token window (Python's `MAX_PROMPT_TOKENS`).
///  - **degeneracy guard**: WhisperKit's built-in temperature fallback performs the retry
///    (`compressionRatioThreshold = 2.4`, the same OpenAI heuristic the Python uses);
///    if a segment is *still* degenerate after fallback we DROP it (return ""), matching
///    the Python's "nothing usable for this segment — skip it".
///
/// The thin transcription seam (the plan's "Transcriber protocol"): `LivePipeline` depends on
/// this, not on WhisperKit, so replay/integration tests can script hypotheses through the REAL
/// pipeline and a non-Whisper engine (e.g. a transducer via sherpa-onnx) can slot in later
/// without touching the pipeline.
protocol Transcribing: Sendable {
    func transcribe(_ audio: [Float], context: String?) async throws -> TranscriptionOutput
    /// Prompt-aware entry point. The head/tail split lets the transcriber budget the priming and the
    /// history from opposite ends (see `ATCContext.PromptParts`). Defaulted so the string-based
    /// engines and the scripted test doubles keep working unchanged.
    func transcribe(_ audio: [Float], prompt: ATCContext.PromptParts?) async throws -> TranscriptionOutput
}

extension Transcribing {
    func transcribe(_ audio: [Float], prompt: ATCContext.PromptParts?) async throws -> TranscriptionOutput {
        try await transcribe(audio, context: prompt?.joined)
    }
}

/// Audio is expected already preprocessed (mono 16 kHz float32 in [-1, 1]); the
/// radio-cleanup stage (`AudioPreprocessor`, ported separately) runs upstream.
actor ATCTranscriber: Transcribing {
    /// Whisper shares a 448-token decoder window between prompt and generated text; cap
    /// the prompt well below it so generation always has room. (= Python `MAX_PROMPT_TOKENS`)
    /// Derived from WhisperKit's own constant, NOT hard-coded: `TextDecoder` trims the prompt with
    /// `suffix(Constants.maxTokenContext / 2 - 1)`, so anything larger loses its front inside
    /// WhisperKit. Deriving it means a dependency bump can't silently re-open that bug.
    static let maxPromptTokens = Constants.maxTokenContext / 2 - 1   // = 111

    private let modelFolder: String
    private let language: String
    private let compressionRatioThreshold: Float
    private let temperatureFallbackCount: Int
    private let cpuOnly: Bool
    private var pipe: WhisperKit?

    /// - Parameter cpuOnly: force CPU compute units (the iOS Simulator has no Neural
    ///   Engine). Leave false on real devices to use the ANE.
    init(modelFolder: String,
         language: String = "en",
         compressionRatioThreshold: Float = 2.4,
         temperatureFallbackCount: Int = 5,
         cpuOnly: Bool = false) {
        self.modelFolder = modelFolder
        self.language = language
        self.compressionRatioThreshold = compressionRatioThreshold
        self.temperatureFallbackCount = temperatureFallbackCount
        self.cpuOnly = cpuOnly
    }

    var isLoaded: Bool { pipe != nil }

    /// Load the converted CoreML model from a local folder. No network (`download: false`).
    /// Mirrors the model load in `ATCTranscriber.__init__`.
    func load() async throws {
        // A superseded model switch cancels its load task, but cancellation cannot interrupt the
        // WhisperKit/CoreML compile below once started — bail HERE so an already-superseded load
        // never begins the multi-GB compile at all (the caller's generation guard handles the
        // supersede-after-start case once the compile returns).
        try Task.checkCancellation()
        let compute = cpuOnly
            ? ModelComputeOptions(melCompute: .cpuOnly, audioEncoderCompute: .cpuOnly, textDecoderCompute: .cpuOnly)
            : nil
        let config = WhisperKitConfig(
            modelFolder: modelFolder,
            computeOptions: compute,
            verbose: false,
            logLevel: .error,
            prewarm: true,
            load: true,
            download: false
        )
        pipe = try await WhisperKit(config)
    }

    /// Transcribe mono 16 kHz audio with an optional context prompt. Returns the transcript +
    /// the ASR confidence; `text` is "" when the decode stays degenerate after fallback (the
    /// caller treats "" as "skip this segment"). Port of `ATCTranscriber.transcribe`.
    func transcribe(_ audio: [Float], context: String? = nil) async throws -> TranscriptionOutput {
        // A bare string has no head/tail structure, so treat it all as head (priming-priority):
        // truncating a caller-supplied prompt from the front is never what they meant.
        try await transcribe(audio, prompt: (context?.isEmpty ?? true)
                             ? nil : ATCContext.PromptParts(head: context!, tail: ""))
    }

    func transcribe(_ audio: [Float], prompt: ATCContext.PromptParts?) async throws -> TranscriptionOutput {
        guard let pipe else { throw TranscriberError.notLoaded }

        let options = DecodingOptions(
            task: .transcribe,
            language: language,                              // pin en; no auto-detect drift
            temperature: 0.0,                                // first pass greedy (clean-audio WER unchanged)
            temperatureFallbackCount: temperatureFallbackCount,  // retries with rising temp on degeneracy
            usePrefillPrompt: true,
            skipSpecialTokens: true,
            withoutTimestamps: true,
            promptTokens: promptTokens(for: prompt, tokenizer: pipe.tokenizer),
            compressionRatioThreshold: compressionRatioThreshold,
            noSpeechThreshold: 0.6
        )

        let results = try await pipe.transcribe(audioArray: audio, decodeOptions: options)

        // Aggregate Whisper's per-segment confidence: mean avgLogprob, max compressionRatio.
        let segments = results.flatMap(\.segments)
        let compression = segments.map(\.compressionRatio).max() ?? 0
        let avgLogprob = segments.isEmpty ? 0 : segments.map(\.avgLogprob).reduce(0, +) / Float(segments.count)
        let asr = ASRConfidence(avgLogprob: avgLogprob, compressionRatio: compression)

        // If still degenerate after WhisperKit's temperature fallback (a stuck repetition loop —
        // "runway three right runway three right ...", which gzip-compresses far better than speech),
        // QW4: try ONE no-prompt re-decode before giving up. Repetition loops are frequently prompt-
        // or greedy-induced, so dropping the context prompt + starting warmer often recovers real
        // text; only if THAT is still degenerate do we drop the segment (return "").
        if compression > compressionRatioThreshold {
            if let recovered = try? await pipe.transcribe(audioArray: audio, decodeOptions: DecodingOptions(
                task: .transcribe, language: language,
                temperature: 0.4, temperatureFallbackCount: temperatureFallbackCount,
                usePrefillPrompt: true, skipSpecialTokens: true, withoutTimestamps: true,
                promptTokens: nil, compressionRatioThreshold: compressionRatioThreshold, noSpeechThreshold: 0.6)) {
                let rSegs = recovered.flatMap(\.segments)
                let rComp = rSegs.map(\.compressionRatio).max() ?? 0
                if rComp <= compressionRatioThreshold {
                    let rText = recovered.map(\.text).joined(separator: " ").trimmingCharacters(in: .whitespacesAndNewlines)
                    let rLp = rSegs.isEmpty ? 0 : rSegs.map(\.avgLogprob).reduce(0, +) / Float(rSegs.count)
                    if !rText.isEmpty { return TranscriptionOutput(text: rText, asr: ASRConfidence(avgLogprob: rLp, compressionRatio: rComp)) }
                }
            }
            return TranscriptionOutput(text: "", asr: asr)
        }
        let text = results.map(\.text).joined(separator: " ")
            .trimmingCharacters(in: .whitespacesAndNewlines)
        return TranscriptionOutput(text: text, asr: asr)
    }

    /// Encode the context to prompt token ids, drop special tokens, and fit the decoder's budget.
    ///
    /// The budget is NOT arbitrary: WhisperKit re-trims with `promptTokens.suffix(maxTokenContext/2 - 1)`
    /// (TextDecoder.swift). Any prompt longer than that loses its FRONT inside WhisperKit — which used
    /// to silently delete every priming section (facility, ownship, plate fixes, and the live ADS-B
    /// "Aircraft on frequency" bias) and forward only rolling history. Budgeting to exactly that size
    /// here makes WhisperKit's `suffix()` a no-op, so what we prioritize is what the model actually sees.
    private func promptTokens(for parts: ATCContext.PromptParts?,
                              tokenizer: WhisperTokenizer?) -> [Int]? {
        guard let parts, !parts.isEmpty, let tokenizer else { return nil }
        let encode = { (s: String) -> [Int] in
            s.isEmpty ? [] : tokenizer.encode(text: " " + s.trimmingCharacters(in: .whitespacesAndNewlines))
                .filter { $0 < tokenizer.specialTokens.specialTokenBegin }
        }
        let ids = Self.assemblePromptIds(head: encode(parts.head), tail: encode(parts.tail),
                                         budget: Self.maxPromptTokens)
        return ids.isEmpty ? nil : ids
    }

    /// Fit `head` + `tail` into `budget` tokens: the head keeps its FRONT (priming order is
    /// priority order), and whatever budget survives is filled with the END of the tail (the most
    /// recent transmissions). Pure + tokenizer-free so the budgeting is unit-testable on its own.
    static func assemblePromptIds(head: [Int], tail: [Int], budget: Int) -> [Int] {
        guard budget > 0 else { return [] }
        let h = head.count > budget ? Array(head.prefix(budget)) : head
        let remaining = budget - h.count
        let t = remaining <= 0 ? [] : (tail.count > remaining ? Array(tail.suffix(remaining)) : tail)
        let out = h + t
        assert(out.count <= budget, "prompt budget overflow — WhisperKit would re-trim from the front")
        return out
    }

    enum TranscriberError: Error { case notLoaded }
}

File: python-legacy/atc_transcriber.py
"""
Lightweight fine-tuned Whisper transcriber for live ATC with optional context prompt.
"""

from __future__ import annotations

import gzip
import os
from typing import Optional

import numpy as np
import torch
from transformers import WhisperForConditionalGeneration, WhisperProcessor

from audio_preprocessing import AudioPreprocessor

# Whisper shares a single 448-token decoder window between the prompt and the
# generated text. If the prompt fills it, generation has no room and transformers
# raises ("max_new_tokens is 0"). Cap the prompt well below 448 so every segment
# always has room to decode, regardless of how long a context the caller passes.
MAX_PROMPT_TOKENS = 220


def _compression_ratio(text: str) -> float:
    """gzip compression ratio of `text` — Whisper's degeneracy signal.

    A repetition loop ("runway three right runway three right ...") compresses
    far better than natural speech, so a high ratio is a reliable tell. This is
    the same heuristic OpenAI's reference decoder uses (default threshold 2.4):
    normal ATC transmissions sit around 1.0–1.8; a stuck loop is well above 2.4.
    Returns 0.0 for empty/very short text (nothing to judge).
    """
    if not text:
        return 0.0
    data = text.encode("utf-8")
    if len(data) < 16:  # too short for the ratio to mean anything
        return 0.0
    return len(data) / max(1, len(gzip.compress(data)))


def _resolve_device(device: str) -> str:
    """Resolve 'auto' to the best available backend: CUDA, then Apple MPS, then CPU."""
    if device != "auto":
        return device
    if torch.cuda.is_available():
        return "cuda"
    if getattr(torch.backends, "mps", None) is not None and torch.backends.mps.is_available():
        return "mps"
    return "cpu"


class ATCTranscriber:
    """Fine-tuned Whisper inference for live ATC segments."""

    def __init__(
        self,
        model_path: str = "models/whisper-atc",
        device: str = "auto",
        enable_preprocessing: bool = True,
        aggressive_preprocessing: bool = True,
        language: str = "en",
        anti_repetition: bool = True,
        compression_ratio_threshold: float = 2.4,
        retry_temperature: float = 0.4,
        retry_repetition_penalty: float = 1.3,
        retry_no_repeat_ngram_size: int = 4,
    ):
        self.device = _resolve_device(device)
        self.model_path = model_path
        # Decode-quality controls. The first decode pass is left exactly as before
        # (greedy, no penalties) so clean-audio accuracy/WER is unchanged; these
        # only govern language pinning and the degenerate-decode retry below.
        self.language = language
        self.anti_repetition = anti_repetition
        self.compression_ratio_threshold = float(compression_ratio_threshold)
        self.retry_temperature = float(retry_temperature)
        self.retry_repetition_penalty = float(retry_repetition_penalty)
        self.retry_no_repeat_ngram_size = int(retry_no_repeat_ngram_size)
        # Metal (MPS) lacks a few ops Whisper uses; fall back to CPU per-op
        # instead of crashing. Keep weights in float32 — fp16 is unstable on MPS.
        if self.device == "mps":
            os.environ.setdefault("PYTORCH_ENABLE_MPS_FALLBACK", "1")

        print(f"Using device: {self.device}")
        print(f"Loading model: {model_path} ...")
        self.processor = WhisperProcessor.from_pretrained(
            model_path, language=language, task="transcribe"
        )
        self.model = WhisperForConditionalGeneration.from_pretrained(model_path)
        # generate() reads forced_decoder_ids from generation_config, NOT model.config.
        # The fine-tuned config ships forced_decoder_ids with the LANGUAGE slot left
        # null (auto-detect), which lets noisy audio decode as another language. Clear
        # it on BOTH so we can pin language=en/task=transcribe per call (see transcribe).
        self.model.config.forced_decoder_ids = None
        self.model.config.suppress_tokens = []
        self.model.generation_config.forced_decoder_ids = None
        self.model = self.model.to(self.device)
        self.model.eval()
        print("[OK] Model loaded")

        self.preprocessor = (
            AudioPreprocessor(sample_rate=16000, aggressive_radio=aggressive_preprocessing)
            if enable_preprocessing
            else None
        )

    def transcribe(
        self,
        audio: np.ndarray,
        context: Optional[str] = None,
    ) -> str:
        """Transcribe mono 16 kHz audio with optional context prompt.

        Returns the transcript, or "" when the decode is degenerate (a stuck
        repetition loop) and a retry can't recover it — the caller should treat
        an empty result as "nothing usable for this segment" and skip it.
        """
        if self.preprocessor is not None:
            audio = self.preprocessor.preprocess(audio)

        input_features = self.processor(
            audio,
            sampling_rate=16000,
            return_tensors="pt",
        ).input_features.to(self.device)

        # Pin the language/task per call. Pinning is necessary because we cleared
        # forced_decoder_ids (see __init__): without it, generate() would
        # auto-detect the language and drift to non-English on low-SNR audio.
        base_kwargs: dict = {"language": self.language, "task": "transcribe"}
        context = (context or "").strip()
        if context:
            # get_prompt_ids returns a bare tensor on recent transformers (5.x)
            # and a BatchEncoding with .input_ids on older ones. Handle both.
            prompt_ids = self.processor.get_prompt_ids(context, return_tensors="pt")
            if hasattr(prompt_ids, "input_ids"):
                prompt_ids = prompt_ids.input_ids
            # Keep the prompt within budget so decoding always has room (see above).
            if prompt_ids.shape[-1] > MAX_PROMPT_TOKENS:
                prompt_ids = prompt_ids[..., :MAX_PROMPT_TOKENS]
            base_kwargs["prompt_ids"] = prompt_ids.to(self.device)

        # First pass: unchanged greedy decode (keeps clean-audio WER identical).
        text = self._generate(input_features, base_kwargs)

        # Degenerate-decode guard. Bare greedy Whisper on clipped/noisy radio can
        # lock into a repetition loop ("runway three right" x60). Such output
        # gzip-compresses far better than speech, so if it trips the threshold we
        # re-roll ONCE with sampling + anti-repetition penalties; if it's still
        # degenerate we drop it (return "") rather than emit the garbage.
        if self.anti_repetition and _compression_ratio(text) > self.compression_ratio_threshold:
            retry_kwargs = dict(
                base_kwargs,
                do_sample=True,
                temperature=self.retry_temperature,
                repetition_penalty=self.retry_repetition_penalty,
                no_repeat_ngram_size=self.retry_no_repeat_ngram_size,
            )
            retry = self._generate(input_features, retry_kwargs)
            text = retry if _compression_ratio(retry) <= self.compression_ratio_threshold else ""

        return text

    def _generate(self, input_features, generate_kwargs: dict) -> str:
        with torch.no_grad():
            predicted_ids = self.model.generate(input_features, **generate_kwargs)
        return self.processor.batch_decode(
            predicted_ids, skip_special_tokens=True
        )[0].strip()
